### PR TITLE
Create sitemap index

### PR DESCRIPTION
**Pull request by kevinmeijer97**
This PR introduces a sitemap.xml route that generates a sitemap index.

We retrieve all relevant sitemaps from Magento based on the current store ID using the sitemap database table. Magento’s sitemaps may include either a urlset (containing links to categories, products, CMS pages, etc.) or another sitemap index, depending on the configured "Maximum No of URLs Per File" setting in Magento.

If a sitemap index is returned, we extract and include the URLs from each referenced urlset in our sitemap index.

You can also add custom indexes to the sitemap by using the rapidez.sitemap.index Eventy filter.

### CODE REVIEW

File: src/Http/Controllers/SitemapController.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+class SitemapController
+{
+    public function index(): Response
+    {
+        // Retrieve cached sitemaps for the current store
+        $sitemaps = Sitemap::getCachedByStoreId();
+
+        // Allow additional sitemaps to be added via the Eventy filter
+        $sitemaps = Eventy::filter('rapidez.sitemap.index', $sitemaps); // @phpstan-ignore-line
```
Comment (BobWez98) [inline src/Http/Controllers/SitemapController.php]: So, the sitemaps pushed in this filter will be on every site? I think we need to specify the store here somehow.

File: src/Models/Sitemap.php [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+class Sitemap extends Model
+{
+    protected $table = 'sitemap';
+
+    public static function getCachedByStoreId(): ?array
+    {
+        $cacheKey = 'sitemaps.'.config('rapidez.store');
+
+        return Cache::rememberForever($cacheKey, function () {
+            return self::where('store_id', config('rapidez.store'))
```
Comment (BobWez98) [inline src/Models/Sitemap.php]: We can use https://github.com/rapidez/core/blob/master/src/Models/Scopes/ForCurrentStoreScope.php for this check.

File: src/Models/Sitemap.php [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+class Sitemap extends Model
+{
+    protected $table = 'sitemap';
+
+    public static function getCachedByStoreId(): ?array
+    {
+        $cacheKey = 'sitemaps.'.config('rapidez.store');
+
+        return Cache::rememberForever($cacheKey, function () {
```
Comment (BobWez98) [inline src/Models/Sitemap.php]: Is remembering forever the way here? When a sitemap in Magento changes, it does not change in the sitemap publicly available until the cache is completely cleared

File: src/Models/Sitemap.php [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+            return self::where('store_id', config('rapidez.store'))
+                ->get()
+                ->flatMap(fn ($sitemap) => self::getSitemapsFromIndex($sitemap->toArray()))
+                ->toArray();
+        });
+    }
+
+    protected static function getSitemapsFromIndex(array $sitemap): array
+    {
+        $sitemapUrl = config('rapidez.magento_url').$sitemap['sitemap_path'].$sitemap['sitemap_filename'];
```
Comment (BobWez98) [inline src/Models/Sitemap.php:37]: If we know the url, why do we need to load the content in? We can just reference it in our index right?

File: src/Models/Sitemap.php [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+                throw new \RuntimeException('Failed to load sitemap XML.');
+            }
+
+            $sitemapOutput = json_decode(json_encode($sitemapXml) ?: '', true);
+
+            return $sitemapOutput['sitemap'] ?? [
+                ['loc' => $sitemapUrl, 'lastmod' => $sitemap['sitemap_time']],
+            ];
+        } catch (\Exception $e) {
+            \Log::error("Error loading sitemap from URL: $sitemapUrl", ['exception' => $e]);
```
Comment (BobWez98) [inline src/Models/Sitemap.php]: We can import this instead of using `\Log`

File: src/Models/Sitemap.php [hunk 6]
```diff
@@ -0,0 +1,47 @@
...
+
+    protected static function getSitemapsFromIndex(array $sitemap): array
+    {
+        $sitemapUrl = config('rapidez.magento_url').$sitemap['sitemap_path'].$sitemap['sitemap_filename'];
+
+        try {
+            $sitemapXml = simplexml_load_file($sitemapUrl, SimpleXMLElement::class, LIBXML_NOERROR | LIBXML_NOWARNING);
+
+            if ($sitemapXml === false) {
+                throw new \RuntimeException('Failed to load sitemap XML.');
```
Comment (BobWez98) [inline src/Models/Sitemap.php]: We can import this instead of using `\RuntimeException`

File: src/Models/Sitemap.php [hunk 7]
```diff
@@ -0,0 +1,47 @@
...
+            }
+
+            $sitemapOutput = json_decode(json_encode($sitemapXml) ?: '', true);
+
+            return $sitemapOutput['sitemap'] ?? [
+                ['loc' => $sitemapUrl, 'lastmod' => $sitemap['sitemap_time']],
+            ];
+        } catch (\Exception $e) {
```
Comment (BobWez98) [inline src/Models/Sitemap.php]: Import

File: src/Models/Sitemap.php [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+        return Cache::rememberForever($cacheKey, function () {
+            return self::where('store_id', config('rapidez.store'))
+                ->get()
+                ->flatMap(fn ($sitemap) => self::getSitemapsFromIndex($sitemap->toArray()))
+                ->toArray();
+        });
+    }
+
+    protected static function getSitemapsFromIndex(array $sitemap): array
```
Comment (BobWez98) [inline src/Models/Sitemap.php:35]: Don't we need to make this function recursive? What if the magento sitemap already is an index?

File: README.md [hunk 9]
```diff
@@ -1,38 +1,42 @@
...
+use TorMorten\Eventy\Facades\Eventy;
+
+Eventy::addFilter('rapidez.sitemap.index', function ($sitemaps) {
+    // Add your custom sitemap URL here
+    $sitemaps[] = [
+        'loc' => url('/some-dynamic-url.xml'),
+        'lastmod' => now()->toDateTimeString(),
+    ];
+
+    return $sitemaps;
```
Comment (BobWez98) [inline README.md:34]: Does this get added to all stores?

**Review by BobWez98 — CHANGES_REQUESTED** [review]

File: src/Http/Controllers/SitemapController.php (hunk 1 above)
Comment (kevinmeijer97) [reply src/Http/Controllers/SitemapController.php]: This is now fixed by setting up a `rapidez.sitemap.{storeId}` filter.

File: README.md (hunk 9 above)
Comment (kevinmeijer97) [reply README.md:34]: Yes, changed it to a filter per store

File: src/Models/Sitemap.php (hunk 4 above)
Comment (kevinmeijer97) [reply src/Models/Sitemap.php:37]: It already is, like i said in the description https://github.com/rapidez/sitemap/pull/1#issue-2632172308

We have to check every sitemap from Magento to see if it is an XML containing just a `<urlset`, if so we add the url of this XML to our index.
The Magento XML can also contain a sitemap index, which would look something like this:

```
<?xml version="1.0" encoding="UTF-8"?>
<sitemapindex xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">
<sitemap><loc>https://test.xyz.nl/media/sitemap/theme/sitemap-2-1.xml</loc><lastmod>2024-11-01T08:10:35+00:00</lastmod></sitemap>
<sitemap><loc>https://test.xyz.nl/media/sitemap/theme/sitemap-2-2.xml</loc><lastmod>2024-11-01T08:10:35+00:00</lastmod></sitemap>
<sitemap><loc>https://test.xyz.nl/media/sitemap/theme/sitemap-2-3.xml</loc><lastmod>2024-11-01T08:10:35+00:00</lastmod></sitemap>
</sitemapindex>
```

In this case we extract each sitemap and add it to our own index.

File: src/Models/Sitemap.php (hunk 8 above)
Comment (kevinmeijer97) [reply src/Models/Sitemap.php:35]: We already do, see https://github.com/rapidez/sitemap/pull/1#discussion_r1827666292.
We do this in the return of line 48.

File: src/Models/Sitemap.php (hunk 3 above)
Comment (kevinmeijer97) [reply src/Models/Sitemap.php]: Changed this to remember it per day

File: src/Http/Controllers/SitemapController.php [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+
+namespace Rapidez\Sitemap\Http\Controllers;
+
+use Illuminate\Http\Response;
+use Rapidez\Sitemap\Models\Sitemap;
+use TorMorten\Eventy\Facades\Eventy;
+
+class SitemapController
+{
+    public function index(): Response
```
Comment (royduin) [inline src/Http/Controllers/SitemapController.php:11]: You *could* just make it invokable that way you don't have to specify the method when registering the route.

**Review by royduin — APPROVED** [review]

**Comment by royduin** [comment]
I'll merge it but feedback is still welcome @indykoning

**Pull request merged**